1519: Move homepage to app folder (start next migration)

dmijatovic opened the pull request:
# Migrate homepage(s) to next app folder

Related to #1519
Closes #1574

Changes proposed in this pull request:
* **start using app folder for home page**. Created basic layout and passed all shared contexts (shared data) as used in page folder. _app and _document file are required as long as we use pages folders. Only index.tsx (homepage) is removed from pages folder.
* refactor auth module: split module into files to be used on client or server side (required by next app approach)  
* fix broken tests and create new test for home page in app folder. The files could be colocated when using app folder approach. So we will have test files next to page file in the future
* fix next.config.js svgr definitions. The build process stopped working after use of app folder. The webpack config needed to be improved.
* chore: bump fe version number to v4.0.1
* chore: update e2e test dependecies. Fixed caching of e2e test github action.
* fix: added 'use client' directive to Helmholtz and Imperial homepage. **Note! This is "quick" fix to make homepages work with the current components without refactoring. The app folder allows for all components to make api calls and be rendered server side but this requires refactoring of the code including "shared" components AppHeader and AppFooter (which are also used in pages folder).**
* feat: migrated add software, add project and add news pages from page to app folder
* chore: fixed failing unit and e2e tests
* **security policy changes**: style policy still needs 'unsafe-inline' mainly with new app folder approach. Further improvements/investigation is needed.
   * **script-src policy changes**: 
       * added nonce to matomo script
       * removed 'strict-dynamic'
       * removed monitoringScripts() url 
       * removed 'unsafe-inline' 
       * removed https:
* chore: upgrade next to latest version (v15.4.2)
* **replace nprogress with bprogress**: nprogress does not work with app folder router. After looking at the alternatives I replaced nprogress with [bprogress](https://bprogress.vercel.app/docs/next/installation) which seem to be "modern" replacement that has specific next implementation.
        
How to test:
* `make start` to build and generate data.
* test homepage for rsd, helmholtz, imperial
* test matomo and consent cookies
* test add software page (as rsd admin and user)
* test add project page (as rsd admin and user)
* test add news page (rsd admin only)
* test other functionalities. In general rsd should work as before :-)

## Points of attention:
- **The router** to be used in the app folder is different module (next/navigation). This module can be used in both app and page folder as far as I see/tested. Old router module (next/router) can only be used in the page folder and should be phased out over time.
- **Directives**: server side specific functionality and the client side specific functionality should not be in the same file. You can use directives ('use server' or 'use client') at the top of the file to "force/indicate" where the code should run. If there is no directive at the top of the file the code will be run on both server and client side but only if the functionality is valid for the mode otherwise you will see an error. The server functionalities are: env variables, async api calls, node api. The client specific functionalties are: useState, useEffect, use..., all functions requiring browser api like window, document, location etc.
- **Routes and layouts**:  it is not possible to have same route in app and page folder. The app folder uses specific naming convention for the route. The folder name is the name of the route and the page.tsx file is required to be in the folder for the route to be valid. The page and layout files are rendered server side. If you need custom layout of the page you create 'layout.tsx' file. You can nest/cascade layouts. For more information [see documentation](https://nextjs.org/docs/app/getting-started/layouts-and-pages)
- **Async server components**: app folder supports async server side components that makes api request very easy. [See documentation](https://nextjs.org/docs/app/getting-started/server-and-client-components)
- **Nonce and style policy**: Currently CSP for styles has 'unsafe-inline'. This is required to have all styles working properly. The style could also work with nonce but not all styles injected have nonce. It seems that integration with MUI/Emotion cache and injecting nonce is not fully covered. Further investigation is needed but current approach works properly.
- **Hydration errors might be caused by browser plugins**: I have hydration error in dev mode caused by Chrome plugin Colorzilla (see images below)
<img width="1506" height="1311" alt="image" src="https://github.com/user-attachments/assets/bfa01862-7a4e-4d0f-9d08-392decd593c2" />
Google search confirms this
<img width="924" height="547" alt="image" src="https://github.com/user-attachments/assets/bb2e13b1-2168-4777-8496-41f35a8646c2" />

## Security headers have same grade with new approach
After the changes we still have A+ grade (maximum)

<img width="1334" height="1310" alt="image" src="https://github.com/user-attachments/assets/7229c9ed-9a7e-4db4-9da0-780c089b841a" />

## Load tests before/after show similar performance (no loss/gain)
The performance is slightly lower with new approach. Before on left and after on right side.

<img width="2539" height="1181" alt="image" src="https://github.com/user-attachments/assets/1a6cabd9-bee5-4eb0-a4dc-6c28268e3900" />

## Matomo tracking
It seems that matomo tracking does not work in some situations, high likely due to personal privacy settings . I have tested tracking on various hardware, OS and browser combinations and discovered the following:
- Windows machine: matomo tracking works only on Edge (which I do not use and do not have personal settings), while it does not work in Chrome and Firefox where I do have strong security and tracking rules). 
- Mac machine: matomo tracking works on Safari only, which I do not use a lot and do not have personal settings, while it does not work on Chrome and Firefox where I use/share same profile with strong security and tracking rules. 
- Linux machine: matomo tracking does not work because I only have Chrome and Firefox with same strong security and tracking rules as on the other machines. 

Based on the results of this test I believe that some personal browser settings might be preventing matomo from tracking even if user "approves" the tracking in our matomo modal. Further investigation about this matter is needed.


PR Checklist:

*   [x] Increase version numbers in `docker-compose.yml`
*   [ ] Link to a GitHub issue
*   [ ] Update documentation
*   [x] Tests

ewan-escience reviewed (commented):
Everything still seems to work.

When I run in dev mode and visit the home page or an add page, I get the following error:

```
Hydration failed because the server rendered HTML didn't match the client. As a result this tree will be regenerated on the client. This can happen if a SSR-ed Client Component used:
```

with the following stack trace:

```
createConsoleError@http://localhost:3000/_next/static/chunks/node_modules_next_dist_client_8f19e6fb._.js:882:71
handleConsoleError@http://localhost:3000/_next/static/chunks/node_modules_next_dist_client_8f19e6fb._.js:1058:54
error@http://localhost:3000/_next/static/chunks/node_modules_next_dist_client_8f19e6fb._.js:1223:57
[project]/node_modules/next/dist/compiled/react-dom/cjs/react-dom-client.development.js [app-client] (ecmascript)/emitPendingHydrationWarnings/<@http://localhost:3000/_next/static/chunks/node_modules_next_dist_compiled_2ce9398a._.js:5153:25
runWithFiberInDEV@http://localhost:3000/_next/static/chunks/node_modules_next_dist_compiled_2ce9398a._.js:3073:131
emitPendingHydrationWarnings@http://localhost:3000/_next/static/chunks/node_modules_next_dist_compiled_2ce9398a._.js:5152:30
completeWork@http://localhost:3000/_next/static/chunks/node_modules_next_dist_compiled_2ce9398a._.js:8331:102
runWithFiberInDEV@http://localhost:3000/_next/static/chunks/node_modules_next_dist_compiled_2ce9398a._.js:3073:131
completeUnitOfWork@http://localhost:3000/_next/static/chunks/node_modules_next_dist_compiled_2ce9398a._.js:10308:23
performUnitOfWork@http://localhost:3000/_next/static/chunks/node_modules_next_dist_compiled_2ce9398a._.js:10245:28
workLoopConcurrentByScheduler@http://localhost:3000/_next/static/chunks/node_modules_next_dist_compiled_2ce9398a._.js:10239:75
renderRootConcurrent@http://localhost:3000/_next/static/chunks/node_modules_next_dist_compiled_2ce9398a._.js:10221:71
performWorkOnRoot@http://localhost:3000/_next/static/chunks/node_modules_next_dist_compiled_2ce9398a._.js:9853:176
performWorkOnRootViaSchedulerTask@http://localhost:3000/_next/static/chunks/node_modules_next_dist_compiled_2ce9398a._.js:10826:26
performWorkUntilDeadline@http://localhost:3000/_next/static/chunks/node_modules_next_dist_compiled_2ce9398a._.js:1982:72

body@unknown:0:0
RootLayout@rsc://React/Server/file:///home/ewan/repo/RSD-as-a-service/frontend/.next/server/chunks/ssr/%5Broot-of-the-server%5D__2bdd46ee._.js?33:1969:264
```

And when I visit the home page in dev mode, I get the following warning in the terminal:

```
 ⚠ ./components/home/rsd/config.ts
Invalid page configuration
Page config in components/home/rsd/config.ts is deprecated. Replace `export const config=…` with the following:
Visit https://nextjs.org/docs/app/api-reference/file-conventions/route-segment-config for more information.
```

ewan-escience reviewed (changes requested):
When I open `frontend/utils/contentSecurityPolicy.ts` I get errors in my IDE that exported functions in a file with `'use server'` should always be `async`.

Furthermore, the file `frontend/middleware.ts` seems to duplicate the logic of `frontend/utils/contentSecurityPolicy.ts`?

dmijatovic commented on frontend/utils/getCspHeader.ts:
```diff
@@ -0,0 +1,45 @@
+// default policies
+export const sharedPolicy = `
+  default-src 'self';
+  style-src 'self' 'unsafe-inline';
```
I removed unsage-inline from script but for style we still need it. Further investigation is needed

dmijatovic commented:
@cmeessen and @jcohen02 Can you have a look at this PR and confirm that the changes work on your homepages too? I did basic test but it would be great if you can doble check and confirm that this approach works on your side.

cmeessen reviewed (approved):
I confirm that the changes work with our current home page.

ewan-escience reviewed (approved):
It all seems to work as far as I can tell. Once this is merged, we should try to get to move over the rest relatively quickly.

In `frontend/utils/contentSecurityPolicy.ts`, there is an unused function `monitoringScripts` that should be removed.

dmijatovic committed:
feat: migrate homepage to app folder
feat: migrate add pages to app folder
fix: e2e action caching and failing e2e tests
fix: failing unit tests
chore: upgrade e2e and frontend dependencies
chore: improve CSP rules
chore: replace nprogress with bprogress bar
chore: eslint remove indent rule because it does not work anymore
chore: added current directory fix in next.config.mjs to supres multiple package-lock files warning
